.github/workflows/release.yaml

```diff
@@ -0,0 +1,151 @@
+          branch: release/v${{env.VERSION}}
+          delete-branch: true
+          title: "Release v${{env.VERSION}}"
+          body: |
+            Release prepared for ${{env.VERSION}}
+            
+            Reminder: Update the changelog
+          token: ${{ steps.generate_token.outputs.token }}
+      - name: Slack Notification
+        run: |
+          jq --null-input '{ text: "BufCLI Release v${{env.VERSION}} has started: ${{ steps.cpr.outputs.pull-request-url }}" }' \
+          | curl -sSL -X POST -H 'Content-Type: application/json' -d @- '${{ secrets.SLACK_RELEASE_NOTIFICATION_WEBHOOK }}'
+  verify:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'push' && startsWith(github.ref_name, 'release') }}
+    steps:
+      - name: Checkout repository code
+        uses: actions/checkout@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          fetch-depth: 0
+      - name: Check changelog is modified
+        run: |
+          files=`(git fetch origin main:main) && (git diff --name-only main)`
+          for file in $files; do
+            if [ "$file" = "CHANGELOG.md" ]; then
+              exit 0
+            fi
+          done
+          echo ERROR: CHANGELOG has not been updated
+          exit 1
+  tag:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'pull_request' && github.event.pull_request.merged == true && startsWith(github.event.pull_request.head.ref, 'release') }}
+    steps:
+      - name: Generate token
+        id: generate_token
+        uses: tibdex/github-app-token@v1
+        with:
+          app_id: 251311
+          private_key: ${{ secrets.TOKEN_EXCHANGE_GH_APP_PRIVATE_KEY }}
+          repository: ${{ github.repository }}
+          permissions: >-
+            {"contents": "write"}
+      - name: Set VERSION variable from tag
+        run: |
+          VERSION=${{github.head_ref}}
+          echo "VERSION=${VERSION##*/}" >> $GITHUB_ENV
+      - name: Checkout repository code
+        uses: actions/checkout@v3
+        with:
+          token: ${{ steps.generate_token.outputs.token }}
+          fetch-depth: 0
+      - name: Tag Release
+        run: |
+          git config --global user.password ${{ steps.generate_token.outputs.token }}
+          git tag -d ${{env.VERSION}} 2> /dev/null || echo 'local ref does not exist'
+          git push origin :${{env.VERSION}} 2> /dev/null || echo 'remote ref does not exist'
+          git tag ${{env.VERSION}}
+          git push origin ${{env.VERSION}}
+  perform:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'push' && startsWith(github.ref_name, 'v') }}
+    steps:
+      - name: Generate token
+        id: generate_token
+        uses: tibdex/github-app-token@v1
+        with:
+          app_id: 251311
+          private_key: ${{ secrets.TOKEN_EXCHANGE_GH_APP_PRIVATE_KEY }}
+          repository: ${{ github.repository }}
+          permissions: >-
+            {"contents": "write", "pull_requests": "write"}
```

Comment:
being pretty pedantic but you could extract this into another job and put 

```
needs: generate-token
```
here

Reply:
if i do this, i would need to pipe the resulting key into the `$GITHUB_OUTPUT` which would leak the secret.